### PR TITLE
Query manager: Unify schemas

### PR DESCRIPTION
**Pull request by sirreal**
This PR removes a fair amount of duplication by providing a single `QueryManager` schema definition for use with serialization. This saves maintaining multiple schemas for a single type of data.

However, the format of items in the schema may be customized. A function is added (`withItemsSchema`) to provide a new schema with the `items` schema injected in the appropriate place.

---

Build size is nearly unaffected:
```
Commit: a95a9053cd8857041d2a46d656e1131054075f69
Author: sirreal
At: 2018-03-22T12:15:19Z
Message: Allow string or integer itemKeys
Delta:
chunk  stat_size           parsed_size           gzip_size
build     -634 B  (-0.0%)       -611 B  (-0.0%)      +46 B  (+0.0%)
```

---

## Testing

1. Tests pass?
1. Smoke test. Use `QueryManager` parts of Calypso (themes, terms, posts, activity log…) and ensure that you see no deserialization warnings and serialize/deserialize continues to behave as before.

### CODE REVIEW

File: client/lib/query-manager/schema.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function withItemsSchema( itemsSchema ) {
+	return cloneDeepWith( queryManagerSchema, value => {
+		if ( value === queryManagerSchema.properties.data.properties.items ) {
+			return itemsSchema;
+		}
+	} );
+}
+
+export default deepFreeze( queryManagerSchema );
```
Comment (sirreal) [inline client/lib/query-manager/schema.js]: I think `deepFreeze` at the definition will work with `cloneDeepWith`, which may make the frozen object clearer.

File: client/lib/query-manager/schema.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			},
+		},
+	},
+} );
+
+/**
+ * Get a queryManagerSchema with a customized items schema
+ *
+ * @param  {Object} itemsSchema Schema that will be used for the items
+ * @return {Object}            Customized schema
```
Comment (ockham) [inline client/lib/query-manager/schema.js]: Nitpick: Whitespace (to align `Schema that [...]` and `Customized schema`).

**Comment by ockham** [comment]
Seeing a bunch of these:

image

Having a hard time repro'ing reliably right now...

**Comment by ockham** [comment]
Attributes other than `cost` this is complaining about include: `version`, `download_url`, `trending_rank`, `popularity_rank`, `launch_date`, `theme_uri`, `description`, `tags`, `preview_url`, `descriptionLong`, `template`, `download`, `taxonomies`, `license`, `license_uri`, `supportDocumentation`, `date_launched`, `date_updated`, `screenshots`, `next`, and `download`.

I think these might be coming from theme details fetched from the WP.org endpoint, which we use to obtain additional information on non-WP.com(-but-potentially-WP.org-)themes installed on a JP site. Maybe just drop the `additionalProperties: false` from the theme schema?

**Comment by sirreal** [comment]
Nice spot, I wasn't sure where `cost` would appear but it makes sense with themes 👍 

I'll try to figure out what's going on.

**Comment by sirreal** [comment]
Thanks @ockham, I polished things up a bit (and relaxed the theme item `additionalProperties`).

It would be nice to clean up all that extra data before it even gets into state, but that's another PR for another day 🙂 

This is ready for another look.

**Comment by ockham** [comment]
> It would be nice to clean up all that extra data before it even gets into state, but that's another PR for another day 🙂

Yep! We have a bunch of normalization functions in https://github.com/Automattic/wp-calypso/blob/master/client/state/themes/utils.js which would probably the correct place to remove unwanted attributes.

**Comment by sirreal** [comment]
> which would probably the correct place to remove unwanted attributes.

If this were on the data-layer…
…just sayin' 🙂

**Pull request merged**